Add ES2015 builds

TylorS opened the pull request:
Add ES2015 module tsconfig that builds to `dist/es2015`.
Add ES2015 module fields `module` and `jsnext:main` for compatibility with es2015-aware bundlers.
Adjust `require()` syntax to `import * as` syntax in order to provide es2015 builds.

Closes #45

blakeembrey commented on src/free-style.spec.ts:
```diff
@@ -1,5 +1,5 @@
-import test = require('blue-tape')
-import crypto = require('crypto')
+import * as test from 'blue-tape'
+import * as crypto from 'crypto'
```
This is not the correct way to use TypeScript.

Frikki replied:
> This is not the correct way to use TypeScript.

???

TylorS replied:
Typescript *does* handle this, it is however non-valid ES6 though. Typescript, `tsc` to be more exact, actually suggests using it as such when attempting to build to ES2015 modules and there are `import foo = require()` style imports.

blakeembrey commented:
I'm not sure I see any advantage and it'll just make the build more complicated. Can you provide more information in your issue on the why before I decide whether the maintenance overhead is worth it?

Frikki commented:
@blakeembrey 

> I'm not sure I see any advantage and it'll just make the build more complicated.

How is it "more complicated"? You run `npm run build` as before.

> ... before I decide whether the maintenance overhead is worth it?

What overhead?

blakeembrey commented:
Honestly, I need to know the why. No one has answered that, so why do it? 

1. Download overhead - you expect every package to provide 2x more weight
2. It changed the requires in a way that's not compatible - if that's required, I can not accept this change
3. Backward compatibility - I can never change this again without breaking someone, so before accepting I **need** to know why

TylorS commented on package.json:
```diff
@@ -4,13 +4,15 @@
   "main": "dist/free-style.js",
   "typings": "dist/free-style.d.ts",
+  "module": "dist/es2015/free-style.js",
+  "jsnext:main": "dist/es2015/free-style.d.ts",
```
Oops, this was a mistake. Nice catch!

TylorS commented:
> Download overhead - you expect every package to provide 2x more weight

Yes, this is true. I think as you suggested in #45 we could not build the delcaration files and use extends to minimize the weight.

> It changed the requires to a way that's incompatible - if that's required, I can not accept this change

This is handled by TypeScript, and nothing is actually broken by this. You can see [this gist](https://gist.github.com/TylorS/3a56ac0700f9ab239eb19d71adc210e3) of the commonjs output is 100% backwards compatible.

> Backward compatibility - I can never change this again without breaking someone, so before accepting I need to know why

I hope I explained why in #45

blakeembrey commented:
Can you please change the required back to the old style? Using the import star for CommonJS main functions is a hack I'd rather not use or promote in my projects. I thought there was a plugin for using CommonJS projects in Rollup - that seems more suitable than trying to alter every project.

TylorS commented:
> Can you please change the required back to the old style? 

TypeScript will not create ES2015 builds with the `require` syntax. 

> I thought there was a plugin for using CommonJS projects in Rollup

There is but its also a poor hack unfortunately.

Frikki commented:
:(

blakeembrey commented:
And that's why I can't accept it right now. The easy fix there is to not build the test file as an ES6 then :) Can TypeScript build ES6 as a single file? Otherwise, how about using `dist.es6` instead of moving the existing parts? Not sure what the naming scheme usually follows.

TylorS commented:
Okay, I updated the PR to revert to the `require()` syntax, and the ES2015 build does not build the test file. :) 

> Can TypeScript build ES6 as a single file?

Unfortunately no :/

> Otherwise, how about using dist.es6 instead of moving the existing parts

The current build is 100% untouched. It compiles to `dist/free-style.js` and `dist/free-style.spec.js` with sourceMaps and `.d.ts` files as they were before. The only *new* files are at `dist/es2015` and those are no `dist/es2015/free-style.js` and `dist/es2015/free-style.js.map`. I could even turn off the sourceMap if you'd prefer to avoid that 1 file, but that may be a disservice to some users, I don't know.

TylorS commented:
I can change the folder name to whatever you'd prefer, `dist.es6` or `dist.es2015` or whatever you see best suits your project.

blakeembrey commented:
Sounds great 👍 Either works, or `esdist`, or whatever 😄

TylorS commented:
@blakeembrey I updated the build to output to `dist.es2015`, its removed during the build step as `dist` has been. I updated the `module` and `jsnext:main` to point to the new build folder, and added `dist.es2015` to the `.gitignore`.

Let me know anything else that you see important.

blakeembrey commented on package.json:
```diff
@@ -4,13 +4,15 @@
   "description": "Make CSS easier and more maintainable by using JavaScript",
   "main": "dist/free-style.js",
   "typings": "dist/free-style.d.ts",
+  "module": "dist.es2015/free-style.js",
```
What is `module`?

TylorS replied:
`module` is the field that the JS community seems to be standardizing on, likely to be the name node ends up using. For example: https://github.com/reactjs/redux/blob/e2e5d30741567bbbe64764cbb4d3868c839de5c6/package.json#L6

`jsnext:main` is the field that rollup itself pioneered.

blakeembrey replied:
If it's not needed yet, shouldn't we avoid using it until it is?

TylorS replied:
Yeah, that is fine with me.

TylorS replied:
Fixed!

blakeembrey commented on package.json:
```diff
@@ -4,13 +4,15 @@
   "scripts": {
     "lint": "tslint \"src/**/*.ts\"",
-    "build": "rimraf dist/ && tsc",
+    "build": "rimraf dist/ dist.es2015/ && tsc && tsc -P tsconfig-es2015.json",
```
A little pedantic, but if we're doing `.` do you mind making it `tsconfig.es2015.json`? 😄

TylorS replied:
Definitely.

TylorS replied:
Fixed

TylorS commented:
Okay, updated! :smile:

TylorS commented on package.json:
```diff
@@ -4,13 +4,14 @@
   "typings": "dist/free-style.d.ts",
+  "jsnext:main": "dist.es2015/free-style.js",
   "files": [
     "dist/",
```
Ahh, hard to do everything perfect. Are you working on this or would you like another PR?

blakeembrey replied:
Done, thanks!

blakeembrey commented:
Released with https://github.com/blakeembrey/free-style/releases/tag/v2.2.0

blakeembrey commented:
I would actually love to know how much and what this changes for Rollup users. I read _a lot_ of other reviews of adding this sort of thing to other modules and didn't see much benefits, but since this module is already only a single file I decided it would be negligible to improve the UX for you.

TylorS commented:
I haven't yet measured every one of my dependencies changes yet, but I can say that for [`@motorcycle/dom`](https://github.com/motorcyclejs/dom) I often see over 200 functions excluded from my bundles as just 1 example.
